agent/consul/gateway_locator.go

```diff
@@ -0,0 +1,284 @@
+package consul
+
+import (
+	"math/rand"
+	"sort"
+	"sync"
+
+	"github.com/hashicorp/consul/agent/consul/state"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/ipaddr"
+	"github.com/hashicorp/consul/lib"
+	"github.com/hashicorp/go-hclog"
+	memdb "github.com/hashicorp/go-memdb"
+)
+
+// GatewayLocator assists in selecting an appropriate mesh gateway when wan
+// federation via mesh gateways is enabled.
+//
+// NOTE: this is ONLY for RPC and gossip so we only need to track local/primary
```

Comment:
It's not super clear to me why we'd only need primary and local for gossip - traditionally gossip needs a full mesh of connectivity.

Reply:
This is only keeping track of the gateways that the consul servers will directly need to dial.

- During secondary bootstrap you need to directly dial the primary DC's gateways.
- After secondary bootstrap all requests to through the local DC's gateway which does the job of routing the request to the correct gateway in a remote DC.

Reply:
I clarified the comment.